File: leetcode/35-search-insert-position/main.go
package main

func searchInsert(nums []int, target int) int {
	min := 0
	max := len(nums)
	for min < max {
		mean := (min + max) / 2
		if target <= nums[mean] {
			max = mean
		} else {
			min = mean + 1
		}
	}
	return min
}

func main() {

}

File: leetcode/35-search-insert-position/main.py
"""
    1st approach: lower bound binary search

    Time    O(logn)
    Space   O(1)
    40 ms, faster than 45.55%
"""


class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        left = 0
        right = len(nums)
        while left < right:
            mid = (left + right)/2
            if target <= nums[mid]:
                right = mid
            else:
                left = mid + 1
        return left
